### PR TITLE
Fix - Poor exception handling for missing packages #1368

### PR DESCRIPTION
**Pull request by adamcamp**
https://github.com/pi-hole/pi-hole/issues/1368

**By submitting this pull request, I confirm the following (please check boxes, eg [X]) _Failure to fill the template will close your PR_:**

***Please submit all pull requests against the `development` branch. Failure to do so will delay or deny your request***

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] I have checked that [another pull request](https://github.com/pi-hole/pi-hole/pulls) for this purpose does not exist.
- [X] I have considered, and confirmed that this submission will be valuable to others.
- [X] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [X] I give this submission freely, and claim no ownership to its content.

**How familiar are you with the codebase?:**

3

---
See https://github.com/pi-hole/pi-hole/issues/1368

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

### CODE REVIEW

File: automated install/basic-install.sh [hunk 1]
```diff
@@ -838,17 +838,13 @@ install_dependent_packages() {
   for i in "${argArray1[@]}"; do
     echo -n ":::    Checking for $i..."
     if ${PKG_MANAGER} -q list installed "${i}" &> /dev/null; then
-      echo " installed!"
+      echo " already installed!"
     else
-      echo " added to install list!"
-      installArray+=("${i}")
+      echo " installing!"
+      "${PKG_INSTALL[@]}" "${i}"
```
Comment (DL6ER) [inline automated install/basic-install.sh]: There is a lot of overhead if you run the package manager for every package on its own.

**Comment by DL6ER** [comment]
Please explain why the title of this PR is

> Fix - Poor exception handling for missing packages

It is not clear to me from looking at the changes you suggest

**Comment by adamcamp** [comment]
@DL6ER ,

Please see the associated bug report for a detailed analysis and to understand the purpose of the changes: 
https://github.com/pi-hole/pi-hole/issues/1368

The problem with running the package manager for all the packages at once is that it doesn't give the proper return code. If you install git and procps-ng at the same time then even if procps-ng fails, it will say the entire install was successful and continue as if nothing went wrong.

I see your point that it could potentially slow down the install, but to me the increased reliability of the install outweighs the performance impact.

Perhaps someone can suggest a flag that can be passed to yum to ensure that it aborts the install after any failures, but I'm not aware of any such flag.

**Comment by dschaper** [comment]
So this is `yum` failing by trying to install a package that doesn't exist? I'm trying to see where the successful install of one package masks out the failure of another, but I think the better approach would be to determine if a package exists prior to trying to install it. Is this with CentOS 6 or 7 that you are seeing the failures?

**Comment by adamcamp** [comment]
See the below example for clarification. Note that the first return code is successful, but the second is failed. The test was on CentOS 7, but it could be the same result on any host with yum and custom repos.

Checking if the package exists beforehand would cover the failure scenario when an RPM doesn't exist, but I'm not sure what would happen if the RPM did exist but it failed to install.

```
[adam@localhost ~]$ sudo yum install git fake_package; echo $?
Loaded plugins: fastestmirror, langpacks
Loading mirror speeds from cached hostfile
 * base: mirror.5ninesolutions.com
 * extras: mirror.5ninesolutions.com
 * updates: mirror.5ninesolutions.com
Package git-1.8.3.1-6.el7_2.1.x86_64 already installed and latest version
No package fake_package available.
Nothing to do
0
[adam@localhost ~]$ sudo yum install fake_package; echo $?
Loaded plugins: fastestmirror, langpacks
Loading mirror speeds from cached hostfile
 * base: mirror.5ninesolutions.com
 * extras: mirror.5ninesolutions.com
 * updates: mirror.5ninesolutions.com
No package fake_package available.
Error: Nothing to do
1

```

**Comment by DL6ER** [comment]
The current yum behavior is quite unfriendly for script usage... See [here](https://bugzilla.redhat.com/show_bug.cgi?id=736694) for discussion I found on this.

In the case of `yum` we might loop over all packages with `yum info <package> >> /dev/null 2>&1 || exit` beforehand? I don't think we have to change the behavior for `apt-get` (but I haven't tested that right now).

**Comment by dschaper** [comment]
`apt` uses the same approach that the Ubuntu/Debian installer uses, `debconf-apt-progress` as a parent to the underlying process. So this would just be for the `yum` (not sure if `dnf` has the same behaviour). But I do like the changes to the installed/installing verbiage, but that would need to be used for both blocks so that users for any of the supported OSes get the same display.

**Comment by Mcat12** [comment]
What's the status on this? It looks like some changes are requested?

**Comment by dschaper** [comment]
Yes, changes were requested, but in the comments, not in a commit review.

**Comment by DL6ER** [comment]
@adamcamp Do you have any update for us?

**Comment by PromoFaux** [comment]
Closing to tidy up, as this appears to have gone stale. Can be reopened if @adamcamp wishes to continue work on it!

**Pull request closed without merging**